README
This package contains ellipse drawing routines designed for pairwise
confidence regions, including distorted ellipses for nonlinear
regression regions.  It also includes a routine `plotcorr()` for 
plotting correlation matrices using ellipses.

The examples require the `MASS`, `nls` and `ts` packages.


NEWS.md
## Version 0.5

* The `pairs()` generic and `pairs.profile()` method have both
been deprecated, and replaced with the `pairs_profile()` function.

## Version 0.4.5 

* `center` was ignored in `ellipse.default()`.

## Version 0.4.4 

* Put source on Github.
* Added a `NEWS.md` file to track changes to the package.
* Added `README.Rmd`.
* Added `pkgdown` site.
* Added US spelling `center` to `ellipse.default()` (thanks to
James Curran).

## Version 0.4.3   

- At request of CRAN, fixed some references.

## Version 0.4.2   

- Update maintainer email.

## Version 0.4.1   

- At request of CRAN, added `pairs()` generic function to avoid
clash with `MASS::pairs.profile()`
- other minor cleanups.

## Version 0.3.8   

- Minor fix (no more `\synopsis` tag in help)

## Version 0.3.7   

- Minor fix (clean up DESCRIPTION file)

## Version 0.3.6   

- Bug fix:  `...` was not being passed from methods to `ellipse.default()`.

## Version 0.3.5   

- License clarification

## Version 0.3.4   

- Update to strict test compliance

## Version 0.3.3   

- Nicer colors in `plotcorr()` example (from Gregor Gorjanc)
		   
## Version 0.3.2   

- Remove deprecated `install.R`

## Version 0.3.1  

- Added NAMESPACE, package overview topic
- Allowed `plotcorr()` to plot just upper or lower triangle
of matrix.

## Version 0.2.14  

- Allowed separate colors for each ellipse in `plotcorr()`
and fixed some `.Rd` errors.

## Version 0.2.12  

- Bug fix for missing values in `plotcorr()`, doc error

## Version 0.0.1 

- First CRAN release
		    






		    






R/ellipse.default.R
"ellipse.default" <-
  function (x, scale = c(1, 1), centre = c(0, 0), level = 0.95,
            t = sqrt(qchisq(level, 2)), which = c(1, 2), npoints = 100,
            center = centre, ...)
{
  if(!missing(centre) && !missing(center)){
    warning("Specify centre or center, not both. The value from center will be used.")
  }

  names <- c("x", "y")
  if (is.matrix(x)) {
    xind <- which[1]
    yind <- which[2]
    r <- x[xind, yind]
    if (missing(scale)) {
      scale <- sqrt(c(x[xind, xind], x[yind, yind]))
      if (scale[1] > 0) r <- r/scale[1]
      if (scale[2] > 0) r <- r/scale[2]
    }
    if (!is.null(dimnames(x)[[1]])) 
      names <- dimnames(x)[[1]][c(xind, yind)]
  }
  else r <- x
  r <- min(max(r,-1),1)  # clamp to -1..1, in case of rounding errors
  d <- acos(r)
  a <- seq(0, 2 * pi, len = npoints)
  matrix(c(t * scale[1] * cos(a + d/2) + center[1], t * scale[2] * 
           cos(a - d/2) + center[2]), npoints, 2, dimnames = list(NULL, 
                                                    names))
}


R/pairs.profile.R
pairs <- function(x, ...) {
  if (inherits(x, "profile"))
    .Deprecated("pairs_profile", msg = "ellipse::pairs is deprecated for profile objects.  Use ellipse::pairs_profile instead.")
  UseMethod("pairs")
}

.pairs_default <- function(x, ...)
  graphics::pairs(x, ...)

.pairs_formula <- function(formula, data = NULL, ..., subset, na.action = stats::na.pass) {
  # This code is copied from graphics:::pairs.formula
  # because the non-standard evaluation wouldn't survive
  # the indirect call.
  
  m <- match.call(expand.dots = FALSE)
  if (is.matrix(eval(m$data, parent.frame()))) 
    m$data <- as.data.frame(data)
  m$... <- NULL
  m$na.action <- na.action
  m[[1L]] <- quote(stats::model.frame)
  mf <- eval(m, parent.frame())
  graphics::pairs(mf, ...)
}

pairs_profile <-
  function (x, labels = c(names(x), "Profile tau"), panel = lines, 
            invert = TRUE, plot.tau = TRUE, plot.trace = TRUE, plot.sketch = TRUE, 
            plot.ellipse = FALSE, level = 0.95, ...) 
{
  stopifnot(inherits(x, "profile"))
    
  doaxis <- function(which, dolabel = TRUE) axis(which, labels = dolabel) # outer = TRUE, line = -0.5, labels = dolabel)
  setup <- function(x, y, ...) plot(range(x[!is.na(x)]), 
                                       range(y[!is.na(y)]), type = "n", axes = FALSE, ...)
  if (is.character(panel)) 
    panel <- get(panel, mode = "function")
  n <- length(x)
  if (plot.tau) 
    n <- n + 1
  oldpar <- par("oma", "mar", "cex", "tck", "mgp", "mex", 
                "mfrow")
  oldcex <- par("cex")
  CEX <- oldcex * max(7.7/(2 * n + 3), 0.6)
  par(mfrow = c(n, n), mgp = c(2, 0.8, 0), oma = rep(3, 4), 
      mar = rep(0.5, 4), tck = -0.03/n)
  on.exit({
    par(oldpar)
  })
  par(cex = CEX)
  if (length(labels) < n) 
    labels <- paste(deparse(substitute(x)), "[,", 1:n, "]", 
                    sep = "")
  if (par("pty") == "s") {
    dif <- diff(par("fin"))/2
    if (dif > 0) 
      par(omi = c(dif * n, 0, dif * n, 0) + par("omi"))
    else par(omi = c(0, (-dif) * n, 0, (-dif) * n) + par("omi"))
  }
  alltau <- unlist(lapply(x, function(x) x[[1]]), use.names = FALSE)
  order <- if (invert) 
    1:n
  else n:1
  for (i in order) {
    for (j in 1:n) {
      if (i <= length(x)) {
          icomp <- x[[i]]
	  ipars <- as.matrix(icomp[[2]])
      }
      if (j <= length(x)) {
          jcomp <- x[[j]]
	  jpars <- as.matrix(jcomp[[2]])
      }
      xx1 <- NA
      xx2 <- NA
      yy1 <- NA
      yy2 <- NA
      if (i <= length(x)) {
        yy1 <- ipars[, i]
        if (j <= length(x)) {
          xx1 <- ipars[, j]
          xx2 <- jpars[, j]
          yy2 <- jpars[, i]
        }
        else {
          xx1 <- icomp[[1]]
        }
      }
      else {
        yy1 <- jcomp[[1]]
        if (j <= length(x)) {
          xx1 <- jpars[, j]
        }
      }
      xx <- c(xx1, NA, xx2)
      yy <- c(yy1, NA, yy2)
      if (i <= length(x)) {
        if (j <= length(x)) 
          setup(xx, yy, ...)
        else setup(alltau, yy, ...)
      }
      else {
        if (j <= length(x)) 
          setup(xx, alltau, ...)
        else setup(alltau, alltau)
      }
      box()
      if (i == 1) 
        doaxis(3, j%%2 == 0)
      if (i == n) 
        doaxis(1, j%%2 == 1)
      if (j == 1) 
        doaxis(2, i%%2 == 0)
      if (j == n) 
        doaxis(4, i%%2 == 1)
      if (i != j) {
        if ((i <= length(x)) && (j <= length(x))) {
          if (plot.trace) 
            panel(xx, yy, ...)
          if (plot.sketch) 
            for (l in level) panel(ellipse(x, which = c(j, 
                                                i), level = l), ...)
          if (plot.ellipse && !is.null(fit <- attr(x, 
                                                   "original.fit"))) 
            for (l in level) panel(ellipse(fit, which = c(j, 
                                                  i), level = l), ...)
        }
        else if (plot.tau) 
          panel(xx, yy, ...)
      }
      else {
        par(usr = c(0, 1, 0, 1))
        text(0.5, 0.5, labels[i], cex = 1.5 * CEX)
      }
    }
  }
  invisible()
}

pairs.profile <- pairs_profile